### PR TITLE
Added user language setting

### PR DESCRIPTION
**Pull request by kolaente**
This pr adds the possibility for users to set their language in the profile settings. When logging in, this setting overwrites the current language setting.

This feature is needed to send [translatable emails](https://github.com/go-gitea/gitea/issues/3622) in the future.

~~Blocked by go-gitea/go-sdk#104~~

### CODE REVIEW

File: routers/user/setting.go [hunk 1]
```diff
@@ -25,6 +25,7 @@ import (
 	"code.gitea.io/gitea/modules/context"
 	"code.gitea.io/gitea/modules/log"
 	"code.gitea.io/gitea/modules/setting"
+	"github.com/Unknwon/i18n"
```
Comment (jonasfranz) [inline routers/user/setting.go]: Fix import order

File: models/migrations/v62.go [hunk 2]
```diff
@@ -0,0 +1,22 @@
+// Copyright 2018 The Gitea Authors. All rights reserved.
+// Use of this source code is governed by a MIT-style
+// license that can be found in the LICENSE file.
+
+package migrations
+
+import (
+	"fmt"
+	"github.com/go-xorm/xorm"
```
Comment (jonasfranz) [inline models/migrations/v62.go]: Fix import order

File: models/migrations/migrations.go [hunk 3]
```diff
@@ -176,6 +176,8 @@ var migrations = []Migration{
 	NewMigration("add is_fsck_enabled column for repos", addFsckEnabledToRepo),
 	// v61 -> v62
 	NewMigration("add size column for attachments", addSizeToAttachment),
+	// v62 -> v63
+	NewMigration("add language comumn for user setting", addLanguageSetting),
```
Comment (jonasfranz) [inline models/migrations/migrations.go]: Fix typo

File: templates/user/settings/profile.tmpl [hunk 4]
```diff
@@ -40,6 +40,20 @@
...
 
+					<div class="field">
+						<label for="language">{{.i18n.Tr "settings.language"}}</label>
+						<div class="ui language selection dropdown" id="language">
+							<input name="language" type="hidden">
+							<i class="dropdown icon"></i>
+							<div class="text">{{.LangName}}</div>
+							<div class="menu">
+							{{range .AllLangs}}
+								<div class="item{{if eq $.Lang .Lang}} active selected{{end}}" data-value="{{.Lang}}">{{.Name}}</div>
```
Comment (jonasfranz) [inline templates/user/settings/profile.tmpl]: You should check if `.Lang` is equal with user's language in database and not with the current language. 
If I set `german` as my language but change it manually by selecting another language (example: `english`) from the dropdown in the footer, the selected language in the form will be `english`. The expected selected language should be `german`.

**Review by jonasfranz — CHANGES_REQUESTED** [review]

File: routers/user/auth.go [hunk 5]
```diff
@@ -333,6 +333,11 @@ func handleSignInFull(ctx *context.Context, u *models.User, remember bool, obeyR
 	ctx.Session.Set("uid", u.ID)
 	ctx.Session.Set("uname", u.Name)
 
+	// Language setting of the user overwrites the one previously set
+	if u.Language != "" {
```
Comment (lafriks) [inline routers/user/auth.go]: len(u.Language) != 0

File: models/migrations/v62.go [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+)
+
+func addLanguageSetting(x *xorm.Engine) error {
+	type User struct {
+		Language string `xorm:"VARCHAR(5)"`
+	}
+
+	if err := x.Sync2(new(User)); err != nil {
```
Comment (jonasfranz) [inline models/migrations/v62.go]: Is it possible that if a user logs in after an upgrade of the database, the language he/she is currently using is set as `language` in database? Do you think that the default language should be `en_US` (at database migration)?

**Comment by jonasfranz** [comment]
@kolaente Please update `go-sdk` in order to fix the drone build.

File: models/migrations/v62.go (hunk 6 above)
Comment (kolaente) [reply models/migrations/v62.go]: I haven't thought of that, I'll implement it.

File: routers/user/setting.go (hunk 1 above)
Comment (kolaente) [reply routers/user/setting.go]: Done.

File: templates/user/settings/profile.tmpl (hunk 4 above)
Comment (kolaente) [reply templates/user/settings/profile.tmpl]: Fixed that.

**Comment by kolaente** [comment]
@JonasFranzDEV done.

File: models/migrations/v62.go (hunk 6 above)
Comment (kolaente) [reply models/migrations/v62.go]: Done.

**Comment by kolaente** [comment]
Does someone know why drone doesn't picks up the right commit?

**Comment by jonasfranz** [comment]
LGTM

**Review by jonasfranz — APPROVED** [review]

**Comment by kolaente** [comment]
Drone fail seems unrelated, the tests work on my machine.

**Review by lafriks — APPROVED** [review]

**Review by lunny — CHANGES_REQUESTED** [review]

File: routers/user/auth.go [hunk 7]
```diff
@@ -339,6 +339,18 @@ func handleSignInFull(ctx *context.Context, u *models.User, remember bool, obeyR
 	ctx.Session.Set("uid", u.ID)
 	ctx.Session.Set("uname", u.Name)
 
+	// Language setting of the user overwrites the one previously set
+	// If the user does not have a locale set, we save the current one.
+	if len(u.Language) == 0 {
+		u.Language = ctx.Locale.Language()
+		if err := models.UpdateUser(u); err != nil {
```
[the remark this answers is not in the document]
Comment (kolaente) [reply routers/user/auth.go]: Fixed that.

**Comment by lafriks** [comment]
@lunny need your approval

**Review by lunny — APPROVED** [review]

**Pull request merged**